### ISSUE TITLE
TrieMap doesn't (always) obey the provided Equiv

### ISSUE THREAD

**Issue by durban** [scala/bug#10481]
`TrieMap`, when used with a custom `Hashing` and `Equiv`, doesn't always find keys which are equivalent according to the custom `Equiv`, but not according to universal equality.

```scala
scala> val h = new scala.util.hashing.Hashing[Int] { def hash(i: Int) = i % 4 }
h: scala.util.hashing.Hashing[Int] = $anon$1@6e909bc8

scala> val e = new scala.math.Equiv[Int] { def equiv(x: Int, y: Int) = (x % 8) == (y % 8) }
e: scala.math.Equiv[Int] = $anon$1@4497204c

scala> val t = new scala.collection.concurrent.TrieMap[Int, String](h, e)
t: scala.collection.concurrent.TrieMap[Int,String] = TrieMap()

scala> t.put(0, "zero")
res15: Option[String] = None

scala> t.get(8)
res16: Option[String] = Some(zero)

scala> t.put(4, "four")
res17: Option[String] = None

scala> t.get(8)
res18: Option[String] = None

scala> t.remove(4)
res19: Option[String] = Some(four)

scala> t.get(8)
res20: Option[String] = Some(zero)
```

I think the reason is that in case of hash collisions, `TrieMap` stores key-value pairs in a `scala.collection.immutable.ListMap` (in the L-nodes), which unconditionally uses universal equality.

/cc @axel22

**Comment by Ichoran** [comment]
That would do it, but I'm not sure that there's a way to fix it within 2.12 and maintain binary compatibility.  I'll make a note to make sure it's fixed in 2.13.

**Comment by axel22** [comment]
@durban That is likely the reason. It should be easily fixable.

**Comment by durban** [comment]
Thanks, good to know that it'll be fixed in 2.13.

Regarding 2.12: isn't `LNode` a private implementation detail of `TrieMap`? How would changing it affect binary compatibility?

**Comment by Ichoran** [comment]
@durban - I think conceptually it's okay if it really is completely private and can't escape, but one would have to think through the issues very carefully by hand because it is private-to-Scala but not private-in-the-JVM. I also I don't know how it might affect serialization.

**Comment by durban** [comment]
Seems to be fixed in 2.13.
